Fix Item model and add ItemFactory + ItemTest for getItem method

anggakahfi opened the pull request:
Menambahkan unit test yang komprehensif untuk method Item::getItem() guna memastikan fungsionalitas pengambilan data item berjalan dengan benar, baik saat ada data maupun saat tabel kosong.

Perubahan Kode yang Dilakukan
1. Perbaikan di Model Item.php

Masalah:
Model belum mendukung Laravel Factory, sehingga unit test tidak dapat memanggil Item::factory().
```php
//Sebelum

<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;
use Exception;
```

```php
//Sesudah
//Menambahkan HasFactory trait agar model Item dapat menggunakan fitur model factory bawaan Laravel untuk keperluan unit testing dan seeding data.
<?php

namespace App\Models;
use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Exception;

class Item extends Model
{
    use HasFactory;
//...
```

2. Menambahkan ItemFactory.php
Masalah: Sebelumnya tidak ada factory sehingga Item::factory()->create() gagal dijalankan.
Solusi: Buat factory lengkap untuk generate data dummy item.

3. Unit Test ItemTest.php
Masalah: SQLite in-memory tidak otomatis mengenali table, sehingga test error no such table: item.
Solusi: Jalankan migration di setUp() sebelum test dijalankan dan pastikan koneksi menggunakan SQLite in-memory.

Test Coverage

PR ini menambahkan 3 kasus pengujian komprehensif:

1.  test_get_item_returns_all_items

Memverifikasi pengambilan semua item berhasil.

Membuat 2 item dummy dan memastikan collection mengembalikan jumlah dan data yang sesuai.

2.  test_get_item_returns_empty_collection_when_no_data

Menguji perilaku ketika tidak ada data di tabel item.

Memastikan method mengembalikan collection kosong (isEmpty()).

3. test_get_item_does_not_throw_error_when_table_empty

Menguji method tetap berjalan normal walaupun tabel kosong.

Memastikan tidak terjadi exception dan collection dikembalikan.

Technical Details

- File Test: tests/Unit/Models/ItemTest.php

- File Model: app/Models/Item.php (getItem() method)

- File Factory: database/factories/ItemFactory.php

Trait: RefreshDatabase untuk isolasi database in-memory

Perintah menjalankan:

php artisan test tests/Unit/Models/ItemTest.php


<img width="812" height="385" alt="image" src="https://github.com/user-attachments/assets/8cdddfee-a80c-42a9-bbe6-8f0179560795" />

anggakahfi committed:
Fix Item model and add ItemFactory + ItemTest for getItem method

anggakahfi commented:
@ardiansyah-sweng

ardiansyah-sweng commented:
perbaiki konfliknya

anggakahfi commented:
baik sudah pak

ardiansyah-sweng commented on tests/Unit/Models/ItemTest.php:
```diff
@@ -2,258 +2,74 @@
-        ]);
+        // Pakai SQLite in-memory
+        config()->set('database.default', 'sqlite');
+        config()->set('database.connections.sqlite.database', ':memory:');
```
databasenya mysql, bukan sqlite.

anggakahfi replied:
sudah diperbaiki pak

tests\Unit\Models\ItemTest.php:

```php
public function test_get_item_returns_all_items()
    {
        // Arrange: buat 3 item dummy
        $items = Item::factory()->count(3)->create([
            'product_id' => 'PROD',
        ]);

        // Act: panggil method getItem()
        $result = (new Item())->getItem();

        // Assert: pastikan hasilnya Collection
        $this->assertInstanceOf(\Illuminate\Database\Eloquent\Collection::class, $result);

        // Assert: jumlah item sama
        $this->assertCount(3, $result);

        // Assert: SKU sama dengan yang dibuat
        $expectedSkus = $items->pluck('sku')->sort()->values();
        $actualSkus = $result->pluck('sku')->sort()->values();
        $this->assertEquals($expectedSkus, $actualSkus);
    }

    /**
     * Test getItem() returns empty collection when no data exists
     */
    public function test_get_item_returns_empty_collection_when_no_data()
    {
        // Pastikan tabel kosong
        Item::query()->delete();

        // Act
        $result = (new Item())->getItem();

        // Assert
        $this->assertInstanceOf(\Illuminate\Database\Eloquent\Collection::class, $result);
        $this->assertTrue($result->isEmpty());
    }

    /**
     * Test getItem() does not throw error when table is empty
     */
    public function test_get_item_does_not_throw_error_when_table_empty()
    {
        // Pastikan tabel kosong
        Item::query()->delete();

        // Act & Assert: pastikan tidak lempar exception
        try {
            $result = (new Item())->getItem();
            $this->assertInstanceOf(\Illuminate\Database\Eloquent\Collection::class, $result);
        } catch (\Throwable $e) {
            $this->fail('getItem() should not throw an exception when table is empty');
        }
    }
```

<img width="782" height="266" alt="image" src="https://github.com/user-attachments/assets/6083b110-15a2-4c12-859b-752266e4310a" />

anggakahfi committed:
Menambah unit test Item::getItem() (pindah dari SQLite ke MySQL) serta update model, factory, dan constant terkait

ardiansyah-sweng commented on app/Models/Item.php:
```diff
@@ -11,31 +12,42 @@
-    
+
     protected $table;
+    protected $fillable = [
```
dihapus saja.

ardiansyah-sweng commented on app/Models/Item.php:
```diff
@@ -11,31 +12,42 @@
-        $this->fillable = ItemColumns::getFillable();
+        // Pakai konfigurasi tabel dari config/db_constants.php
+        $this->table = config('db_constants.table.item', 'items');
+        $this->fillable = array_values(config('db_constants.column.item') ?? $this->fillable);
```
ganti dengan:         $this->fillable = ItemColumns::getFillable();

anggakahfi replied:
sudah diperbaiki pak

ardiansyah-sweng commented on app/Models/Item.php:
```diff
@@ -73,76 +86,77 @@ public static function deleteItemById($id)
         return self::create($data);
     }
 
-    public function unit()
```
jangan dihapus

ardiansyah-sweng commented on app/Models/PurchaseOrderDetail.php:
```diff
@@ -2,13 +2,24 @@
\ No newline at end of file
+
+    // Kolom yang bisa diisi secara mass assignment
+    protected $fillable = [
```
hapus saja

ardiansyah-sweng commented on database/factories/ItemFactory.php:
```diff
@@ -11,28 +11,24 @@
+    // KODE ANDA YANG SUDAH WORKING - TETAP DIPERTAHANKAN
     public function definition()
     {
         return [
```
jangan dihapus

anggakahfi replied:
sudah diperbaiki pak

anggakahfi committed:
fix: Fix every file agar menggunakan konstanta kolom dari ItemColumns.php

anggakahfi committed:
Merge branch 'development' into Anggasta_434A_ItemModel_getItem

anggakahfi commented:
@ardiansyah-sweng

anggakahfi committed:
menambahkan } para fungsi 

test_get_item_does_not_throw_error_when_table_empty

anggakahfi commented:
<img width="661" height="277" alt="image" src="https://github.com/user-attachments/assets/bb458498-21b3-4dce-8db2-c0cea8176b1f" />

ardiansyah-sweng commented:
perbaiki konfliknya

anggakahfi committed:
Menambah POD Columns dan Factoy agar fungsi it throws exception di ItemTest sukses

anggakahfi committed:
Merge branch 'development' into Anggasta_434A_ItemModel_getItem

anggakahfi commented:
> perbaiki konfliknya

sudah diperbaiki pak

ardiansyah-sweng commented on app/Models/Item.php:
```diff
@@ -2,58 +2,68 @@
         parent::__construct($attributes);
 
-        // Use table name from config
-        $this->table = config('db_tables.item', 'items');
```
kenapa ini dihapus?

anggakahfi committed:
Update table configuration key in Item model

Menggunakan config\db_tables.php sebagai referensi instead of config\db_constants.php
